File: LeetCode/1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.java
class Solution {
    private int[] mem;
    private int n;
    public int solve(int i, int[] arr, int k) {
        if (i >= n) {
            return 0;
        }
        if (mem[i] != -1) {
            return mem[i];
        }
        int result = 0, currMax = -1;
        for (int j = i; j < n && j - i + 1 <= k; j++) {
            currMax = Math.max(currMax, arr[j]);
            result = Math.max(result, ((j - i + 1) * currMax) + solve(j + 1, arr, k));
        }
        return mem[i] = result;
    }
    public int maxSumAfterPartitioning(int[] arr, int k) {
        mem = new int[501];
        Arrays.fill(mem, -1);
        n = arr.length;
        return solve(0, arr, k);
    }
}

File: LeetCode/1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.py
class Solution:
    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
        mem = [-1] * 501
        n = len(arr)

        def solve(i: int, arr: List[int], k: int) -> int:
            if (i >= n):
                return 0
            if mem[i] != -1:
                return mem[i]
            cur_max = -1
            result = 0
            for j in range(i, min(n, i + k)):
                cur_max = max(cur_max, arr[j])
                result = max(result, cur_max * (j - i + 1) + solve(j + 1, arr, k))

            mem[i] = result
            return result

        return solve(0, arr, k)